### ISSUE TITLE
S3 provider - Video upload in another folder 

### ISSUE THREAD

**Issue by alejandrohg1** [muxinc/next-video#149]
Hi, i was using amazon s3 , and i was trying to save the videos in a folder instead of the root folder. is there a way to specified the folder in the s3 provider configs. thank you so much 

![image](https://github.com/muxinc/next-video/assets/59666985/bb9007c3-859b-41bc-a1b8-cb7613be4081)

**Comment by DewinU** [comment]
+1, I Have the same problem, looks like in the S3 Provider in the PutObject command the key for will always be just the name of the file (thanks to `path.basename`), so the file will be always be saved on the root of the S3.

Maybe we can a add an optional `folder` parameter in the S3 provider to add a custom folder where to save the videos.

```typescript
async function putAsset(filePath: string, size: number, stream: ReadStream | Readable) {
  log.info(log.label('Uploading file:'), `${filePath} (${size} bytes)`);

  try {
    await putObject(s3, {
      ACL: 'public-read',
      Bucket: bucketName,
      Key: path.basename(filePath),
      Body: stream,
      ContentLength: size,
    });
```

**Comment by luwes** [comment]
yes, that seems like a useful feature! thanks for the feedback.
I'll add this for the bucket based providers. 

`/videos/` will translate to bucket root but folders in `/videos/` will be kept if this is possible.

so `/videos/cats/clip1.mp4` will go to `/cats/clip1.mp4` in the bucket...

would that work for your use cases?

**Comment by alejandrohg1** [comment]
Yes that would be perfect thank you so much

**Comment by DewinU** [comment]
Is this will also work for videos that come from remote sources? like a video from `example.s3.com.video.mp4` i wanted to save this and another remote videos to videos(`root S3`)/todoApp/videos

**Comment by luwes** [comment]
that's a good call. it can't work with remote source urls. 
we could add an extra config option but it would be fixed to one folder, does that work?

**Comment by luwes** [comment]
this is implemented in v0.14.0, would be great if you could test it out!

you can change the path locally to the video config JSON as well with a `remoteSourceAssetPath` callback.


an example next.config.js

```js
// @ts-check
const path = require('path');
const crypto = require('crypto');
const { withNextVideo } = require('next-video/process')

/** @type {import('next').NextConfig} */
const nextConfig = {}

module.exports = withNextVideo(nextConfig, {
  remoteSourceAssetPath,
  provider: 'amazon-s3',
  providerConfig: {
    'amazon-s3': {
      endpoint: 'https://s3.us-east-1.amazonaws.com',
      generateAssetKey
    }
  },
})

function remoteSourceAssetPath(url) {
  const urlObj = new URL(url);
  return 'remote/' + toSafePath(decodeURIComponent(`${urlObj.hostname}${urlObj.pathname}`));
}

function generateAssetKey(filePathOrURL, folder) {
  if (!isRemote(filePathOrURL)) return filePathOrURL;

  const url = new URL(filePathOrURL);
  let hostnameHash = crypto.createHash('md5').update(url.hostname).digest('hex');

  return `${folder}/remote/${hostnameHash}/${path.basename(decodeURIComponent(url.pathname))}`;
}

function isRemote(filePath) {
  return /^https?:\/\//.test(filePath);
}

function toSafePath(str) {
  return str
    .replace(/^[^a-zA-Z0-9]+|[^a-zA-Z0-9]+$/g, '')
    .replace(/[^a-zA-Z0-9._-]+/g, '_');
}
```
